// react-google-places-autocomplete/dist/react-google-places-autocomplete.cjs.prod.js
"use strict";

function _interopDefault(ex) {
  return ex && "object" == typeof ex && "default" in ex ? ex.default : ex;
}

Object.defineProperty(exports, "__esModule", {
  value: !0
});

var React = _interopDefault(require("react")), PropTypes = _interopDefault(require("prop-types")), SCRIPT_ID = "react-google-places-autocomplete", injectScript = function(apiKey) {
  if (!document.getElementById(SCRIPT_ID)) {
    var script = document.createElement("script");
    script.id = SCRIPT_ID, script.type = "text/javascript", script.src = "https://maps.googleapis.com/maps/api/js?key=" + apiKey + "&libraries=places", 
    document.body.appendChild(script);
  }
};

function isNativeReflectConstruct() {
  if ("undefined" == typeof Reflect || !Reflect.construct) return !1;
  if (Reflect.construct.sham) return !1;
  if ("function" == typeof Proxy) return !0;
  try {
    return Date.prototype.toString.call(Reflect.construct(Date, [], (function() {}))), 
    !0;
  } catch (e) {
    return !1;
  }
}

function _construct(Parent, args, Class) {
  return (_construct = isNativeReflectConstruct() ? Reflect.construct : function(Parent, args, Class) {
    var a = [ null ];
    a.push.apply(a, args);
    var instance = new (Function.bind.apply(Parent, a));
    return Class && _setPrototypeOf(instance, Class.prototype), instance;
  }).apply(null, arguments);
}

function _setPrototypeOf(o, p) {
  return (_setPrototypeOf = Object.setPrototypeOf || function(o, p) {
    return o.__proto__ = p, o;
  })(o, p);
}

function _extends() {
  return (_extends = Object.assign || function(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) Object.prototype.hasOwnProperty.call(source, key) && (target[key] = source[key]);
    }
    return target;
  }).apply(this, arguments);
}

var autocompletionRequestBuilder = function(autocompletionRequest) {
  var res = _extends({}, autocompletionRequest);
  return autocompletionRequest.bounds && (res.bounds = _construct(google.maps.LatLngBounds, autocompletionRequest.bounds)), 
  autocompletionRequest.location && (res.location = new google.maps.LatLng(autocompletionRequest.location)), 
  res;
}, debounce = function(callback, timeout) {
  var d, e;
  return function() {
    function helper() {
      d = null, e = callback.apply(thisRef, argumentsRef);
    }
    var thisRef = this, argumentsRef = arguments;
    return clearTimeout(d), !(d = setTimeout(helper, timeout)) && (e = callback.apply(thisRef, argumentsRef)), 
    e;
  };
}, latLngBoundsType = function(props, propName, componentName) {
  var prop = props[propName];
  return prop ? Array.isArray(prop) && 2 === prop.length && prop.every((function(value) {
    return 2 === Object.keys(value).length && value.hasOwnProperty("lat") && value.hasOwnProperty("lng") && Number(value.lat) && Number(value.lng);
  })) ? null : new Error("Invalid prop `" + propName + "` supplied to `" + componentName + "`. Validation failed.") : null;
}, componentRestrictionsType = PropTypes.shape({
  country: PropTypes.oneOfType([ PropTypes.string, PropTypes.arrayOf(PropTypes.string) ])
}), latLngType = PropTypes.shape({
  lat: PropTypes.number,
  lng: PropTypes.number
}), autocompletionRequestType = PropTypes.shape({
  bounds: latLngBoundsType,
  componentRestrictions: componentRestrictionsType,
  location: latLngType,
  offset: PropTypes.number,
  radius: PropTypes.number,
  types: PropTypes.arrayOf(PropTypes.string)
}), suggestionClassNamesType = PropTypes.shape({
  container: PropTypes.string,
  suggestion: PropTypes.string,
  suggestionActive: PropTypes.string
}), suggestionStylesType = PropTypes.shape({
  container: PropTypes.object,
  suggestion: PropTypes.object
});

function _extends$1() {
  return (_extends$1 = Object.assign || function(target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) Object.prototype.hasOwnProperty.call(source, key) && (target[key] = source[key]);
    }
    return target;
  }).apply(this, arguments);
}

function _assertThisInitialized(self) {
  if (void 0 === self) throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return self;
}

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype), subClass.prototype.constructor = subClass, 
  subClass.__proto__ = superClass;
}

function _defineProperty(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value: value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}

var GooglePlacesAutocomplete = function(_React$Component) {
  function GooglePlacesAutocomplete(props) {
    var _this;
    return _defineProperty(_assertThisInitialized(_this = _React$Component.call(this, props) || this), "fetchSuggestions", debounce((function(value) {
      var _this$props = _this.props, autocompletionRequest = _this$props.autocompletionRequest, withSessionToken = _this$props.withSessionToken, sessionToken = _this.state.sessionToken, autocompletionReq = _extends$1({}, autocompletionRequest);
      withSessionToken && sessionToken && (autocompletionReq.sessionToken = sessionToken), 
      _this.setState({
        loading: !0
      }), _this.placesService.getPlacePredictions(_extends$1({}, autocompletionRequestBuilder(autocompletionReq), {
        input: value
      }), _this.fetchSuggestionsCallback);
    }), _this.props.debounce)), _defineProperty(_assertThisInitialized(_this), "initalizeService", (function() {
      if (window.google) if (window.google.maps) if (window.google.maps.places) _this.placesService = new window.google.maps.places.AutocompleteService, 
      _this.setState({
        placesServiceStatus: window.google.maps.places.PlacesServiceStatus.OK,
        timeoutId: null
      }), _this.generateSessionToken(); else {
        console.error("[react-google-places-autocomplete]: Google maps places script not loaded");
        var _timeoutId2 = setTimeout(_this.initializeService, 500);
        _this.setState({
          timeoutId: _timeoutId2
        });
      } else {
        console.error("[react-google-places-autocomplete]: Google maps script not loaded");
        var _timeoutId = setTimeout(_this.initalizeService, 500);
        _this.setState({
          timeoutId: _timeoutId
        });
      } else {
        console.error("[react-google-places-autocomplete]: Google script not loaded");
        var timeoutId = setTimeout(_this.initalizeService, 500);
        _this.setState({
          timeoutId: timeoutId
        });
      }
    })), _defineProperty(_assertThisInitialized(_this), "generateSessionToken", (function() {
      var sessionToken = new google.maps.places.AutocompleteSessionToken;
      _this.setState({
        sessionToken: sessionToken
      });
    })), _defineProperty(_assertThisInitialized(_this), "handleClick", (function(ev) {
      var idPrefix = _this.props.idPrefix;
      ev.target.id.includes(idPrefix + "-google-places-autocomplete") || _this.clearSuggestions();
    })), _defineProperty(_assertThisInitialized(_this), "changeValue", (function(value) {
      _this.setState({
        value: value
      }), value.length > 0 ? _this.fetchSuggestions(value) : _this.setState({
        suggestions: []
      });
    })), _defineProperty(_assertThisInitialized(_this), "onSuggestionSelect", (function(suggestion, ev) {
      void 0 === ev && (ev = null), ev && ev.stopPropagation();
      var _this$props2 = _this.props, displayFromSuggestionSelected = _this$props2.displayFromSuggestionSelected, onSelect = _this$props2.onSelect;
      _this.setState({
        activeSuggestion: null,
        suggestions: [],
        value: displayFromSuggestionSelected(suggestion)
      }), _this.generateSessionToken(), onSelect(suggestion);
    })), _defineProperty(_assertThisInitialized(_this), "fetchSuggestionsCallback", (function(suggestions, status) {
      _this.state.placesServiceStatus;
      _this.setState({
        loading: !1,
        suggestions: suggestions || []
      });
    })), _defineProperty(_assertThisInitialized(_this), "handleKeyDown", (function(event) {
      var _this$state = _this.state, activeSuggestion = _this$state.activeSuggestion, suggestions = _this$state.suggestions;
      switch (event.key) {
       case "Enter":
        event.preventDefault(), null !== activeSuggestion && _this.onSuggestionSelect(suggestions[activeSuggestion]);
        break;

       case "ArrowDown":
        _this.changeActiveSuggestion(1);
        break;

       case "ArrowUp":
        _this.changeActiveSuggestion(-1);
        break;

       case "Escape":
        _this.clearSuggestions();
      }
    })), _defineProperty(_assertThisInitialized(_this), "clearSuggestions", (function() {
      _this.setState({
        activeSuggestion: null,
        suggestions: []
      });
    })), _defineProperty(_assertThisInitialized(_this), "renderInput", (function() {
      var _assertThisInitialize = _assertThisInitialized(_this), value = _assertThisInitialize.state.value, _assertThisInitialize2 = _assertThisInitialize.props, idPrefix = _assertThisInitialize2.idPrefix, inputClassName = _assertThisInitialize2.inputClassName, inputStyle = _assertThisInitialize2.inputStyle, placeholder = _assertThisInitialize2.placeholder, renderInput = _assertThisInitialize2.renderInput, required = _assertThisInitialize2.required, disabled = _assertThisInitialize2.disabled;
      return renderInput ? renderInput({
        autoComplete: "off",
        id: (idPrefix ? idPrefix + "-" : "") + "react-google-places-autocomplete-input",
        value: value,
        onChange: function(_ref) {
          var target = _ref.target;
          return _this.changeValue(target.value);
        },
        onKeyDown: _this.handleKeyDown,
        type: "text",
        placeholder: placeholder,
        required: required,
        disabled: disabled
      }) : React.createElement("input", {
        autoComplete: "off",
        className: inputClassName || "google-places-autocomplete__input",
        id: (idPrefix ? idPrefix + "-" : "") + "react-google-places-autocomplete-input",
        onChange: function(_ref2) {
          var target = _ref2.target;
          return _this.changeValue(target.value);
        },
        onKeyDown: _this.handleKeyDown,
        placeholder: placeholder,
        style: inputStyle,
        type: "text",
        value: value,
        required: required,
        disabled: disabled
      });
    })), _defineProperty(_assertThisInitialized(_this), "renderLoader", (function() {
      var loader = _this.props.loader;
      return loader || React.createElement("div", {
        className: "google-places-autocomplete__suggestions-container"
      }, React.createElement("div", {
        className: "google-places-autocomplete__suggestions"
      }, "Loading..."));
    })), _defineProperty(_assertThisInitialized(_this), "renderSuggestions", (function() {
      var _assertThisInitialize3 = _assertThisInitialized(_this), _assertThisInitialize4 = _assertThisInitialize3.state, activeSuggestion = _assertThisInitialize4.activeSuggestion, suggestions = _assertThisInitialize4.suggestions, _assertThisInitialize5 = _assertThisInitialize3.props, idPrefix = _assertThisInitialize5.idPrefix, renderSuggestions = _assertThisInitialize5.renderSuggestions, suggestionsClassNames = _assertThisInitialize5.suggestionsClassNames, suggestionsStyles = _assertThisInitialize5.suggestionsStyles;
      return 0 === suggestions.length ? null : renderSuggestions ? renderSuggestions(activeSuggestion, suggestions, _this.onSuggestionSelect) : React.createElement("div", {
        id: idPrefix + "-google-places-suggestions-container",
        className: suggestionsClassNames.container || "google-places-autocomplete__suggestions-container",
        style: suggestionsStyles.container
      }, suggestions.map((function(suggestion, index) {
        return React.createElement("div", {
          id: idPrefix + "-google-places-autocomplete-suggestion--" + index,
          key: suggestion.id,
          className: (suggestionsClassNames.suggestion || "google-places-autocomplete__suggestion") + " " + (activeSuggestion === index ? suggestionsClassNames.suggestionActive || "google-places-autocomplete__suggestion--active" : ""),
          style: suggestionsStyles.suggestion,
          onClick: function(event) {
            return _this.onSuggestionSelect(suggestion, event);
          },
          role: "presentation"
        }, suggestion.description);
      })));
    })), _this.state = {
      activeSuggestion: null,
      loading: !1,
      placesServiceStatus: null,
      sessionToken: null,
      suggestions: [],
      timeoutId: null,
      value: props.initialValue
    }, _this;
  }
  _inheritsLoose(GooglePlacesAutocomplete, _React$Component);
  var _proto = GooglePlacesAutocomplete.prototype;
  return _proto.componentDidMount = function() {
    var apiKey = this.props.apiKey;
    apiKey && injectScript(apiKey), this.initalizeService(), document.addEventListener("click", this.handleClick);
  }, _proto.componentWillUnmount = function() {
    var timeoutId = this.state.timeoutId;
    timeoutId && clearTimeout(timeoutId), document.removeEventListener("click", this.handleClick);
  }, _proto.UNSAFE_componentWillReceiveProps = function(nextProps) {
    var initialValue = this.props.initialValue;
    nextProps.initialValue !== initialValue && this.setState({
      value: nextProps.initialValue
    });
  }, _proto.changeActiveSuggestion = function(direction) {
    if (0 !== this.state.suggestions.length) switch (direction) {
     case 1:
      this.setState((function(_ref3) {
        var activeSuggestion = _ref3.activeSuggestion, suggestions = _ref3.suggestions;
        return null === activeSuggestion || activeSuggestion === suggestions.length - 1 ? {
          activeSuggestion: 0
        } : {
          activeSuggestion: activeSuggestion + 1
        };
      }));
      break;

     case -1:
      this.setState((function(_ref4) {
        var activeSuggestion = _ref4.activeSuggestion, suggestions = _ref4.suggestions;
        return activeSuggestion ? {
          activeSuggestion: activeSuggestion - 1
        } : {
          activeSuggestion: suggestions.length - 1
        };
      }));
    }
  }, _proto.render = function() {
    var loading = this.state.loading;
    return React.createElement("div", {
      className: "google-places-autocomplete"
    }, this.renderInput(), loading ? this.renderLoader() : this.renderSuggestions());
  }, GooglePlacesAutocomplete;
}(React.Component);

GooglePlacesAutocomplete.propTypes = {}, GooglePlacesAutocomplete.defaultProps = {
  apiKey: "",
  autocompletionRequest: {},
  debounce: 300,
  disabled: !1,
  idPrefix: "",
  initialValue: "",
  inputClassName: "",
  inputStyle: {},
  loader: null,
  onSelect: function() {},
  displayFromSuggestionSelected: function(suggestion) {
    return suggestion.description;
  },
  placeholder: "Address...",
  renderInput: void 0,
  renderSuggestions: void 0,
  required: !1,
  suggestionsClassNames: {
    container: "",
    suggestion: "",
    suggestionActive: ""
  },
  suggestionsStyles: {
    container: {},
    suggestion: {}
  },
  withSessionToken: !1
};

var geocodeByAddress = function(address) {
  var geocoder = new window.google.maps.Geocoder, OK = window.google.maps.GeocoderStatus.OK;
  return new Promise((function(resolve, reject) {
    geocoder.geocode({
      address: address
    }, (function(results, status) {
      return status !== OK ? reject(status) : resolve(results);
    }));
  }));
}, getLatLng = function(result) {
  return new Promise((function(resolve, reject) {
    try {
      return resolve({
        lat: result.geometry.location.lat(),
        lng: result.geometry.location.lng()
      });
    } catch (e) {
      return reject(e);
    }
  }));
}, geocodeByPlaceId = function(placeId) {
  var geocoder = new window.google.maps.Geocoder, OK = window.google.maps.GeocoderStatus.OK;
  return new Promise((function(resolve, reject) {
    geocoder.geocode({
      placeId: placeId
    }, (function(results, status) {
      return status !== OK ? reject(status) : resolve(results);
    }));
  }));
};

exports.default = GooglePlacesAutocomplete, exports.geocodeByAddress = geocodeByAddress, 
exports.geocodeByPlaceId = geocodeByPlaceId, exports.getLatLng = getLatLng;
